test/fail_compilation/test20245.d

```diff
@@ -0,0 +1,44 @@
+/*
+REQUIRED_ARGS: -preview=dip1000
+TEST_OUTPUT:
+---
+fail_compilation/test20245.d(15): Error: scope variable `a` may not be returned
+fail_compilation/test20245.d(19): Error: cannot take address of `scope` parameter `x` in `@safe` function `foo`
+fail_compilation/test20245.d(25): Error: scope variable `a` may not be returned
+fail_compilation/test20245.d(34): Error: reference to local variable `price` assigned to non-scope `this.minPrice`
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=20245
+@safe int* foo(ref int x) {
+    int* a = &x;
+    return a;
+}
+
+@safe int** foo(ref scope int* x) {
+    int** a = &x;
+    return a;
```

Comment:
This error is inconsistent to the other test cases, `a` should be inferred as `scope`. Seems to be yet another weird special case based on `hasIndirections`?

Reply:
There's two checks: first, in expressionsem.d there's `checkAddressVar` to check whether it is possible at all to take the address safely. This fails because it's taking the address of a `scope` variable with pointers, and there's only one layer of `scope`. This test case is taken from Walter's issue description:
> The following should not compile:
> (...)
> because there's not way to attach `scope` to anything but the head, i.e. no scope pointers to scope pointers to int.

The second check is in escape.d where all the `scope` / `return` is inferred and escaping `scope` variables are detected, which is where the "scope variable `a` may not be returned" comes from.

Reply:
Right, DIP1000 falls apart once there's multiple layers of indirections. Got tricked again by the kinda misleading error message